### PR TITLE
Add logic to make `Playfield` can spawn note from a target time

### PR DESCRIPTION
**Pull request by HelloYeew**
Currently when the note spawn we need to launch the method `SpawnNote` or `SpawnTapNote` in `Playfield` but this PR add a new `Playfield` ability to make `Playfield` can correctly spawn a note at the correct time.

- Now `DrawableNote` has a property call `TargetTime` to collect the time that the note need to be hit or get to the goal.
- In `Playfield` I add a new list call `NotesPool` for the future use too. And if you need to spawn using target time you need to put your note in this list too to make a thing work.
- The old logic are not gone but the old logic use `SPEED` as a note velocity but in the new logic I set the `TIME_NOTE_APPEARS` so the note need to spawn in `note.TargetTime - TIME_NOTE_APPEARS` and must reach in time.

# Testing

This PRs is somehow hard to test on how accurate is this thing work but you can test it by add some `LogPrint` to the note spawn and despawn logic in `Playfield`

```csharp
protected override void Update()
{
    if (NotesPool != null)
    {
        foreach (Drawable note in NotesPool.ToList())
        {
            // We have a lot of note type so we need to check on its type before we can update it
            if (note is DrawableTapNote tapNote)
            {
                if (tapNote.TargetTime != 0f && Clock.TimeInfo.Current >= tapNote.TargetTime - TIME_NOTE_APPEARS)
                {
                    AddInternal(note);
                    Logger.LogPrint("Note spawn at " + Clock.TimeInfo.Current); // Spawn logging
                    NotesPool.Remove(tapNote);
                }
            }
        }
    }

    foreach (Drawable note in InternalChildren.ToList())
    {
        if (note is DrawableTapNote tapNote)
        {
            if (MathUtils.EuclideanDistance(NoteLaneExtension.GetSpawnerPosition(tapNote.Lane), NoteLaneExtension.GetSensorPosition(tapNote.Lane)) <
                MathUtils.EuclideanDistance(tapNote.Position, NoteLaneExtension.GetSpawnerPosition(tapNote.Lane)))
            {
                // Enter despawning state
                note.FadeOut(50, Easing.InBounce);
                Scheduler.AddDelayed(() => RemoveInternal(note), 500);
                Logger.LogPrint("Despawned at " + Clock.TimeInfo.Current); // Despawn logging
                // Don't forget to remove DISTANCE_ON_DESPAWN before test!
            }
            else
            {
                if (tapNote.TargetTime != 0f)
                {
                    double speed = MathUtils.EuclideanDistance(
                        NoteLaneExtension.GetSpawnerPosition(tapNote.Lane),
                        NoteLaneExtension.GetSensorPosition(tapNote.Lane)) / TIME_NOTE_APPEARS;
                    tapNote.Position = new Vector2(
                        (float) (tapNote.Position.X + ((float)(speed * (-(float)Math.Cos((NoteLaneExtension.GetAngle(tapNote.Lane) + 90f) * (float)(Math.PI / 180)))) * Clock.ElapsedFrameTime)),
                        (float) (tapNote.Position.Y + ((float)(speed * (-(float)Math.Sin((NoteLaneExtension.GetAngle(tapNote.Lane) + 90f) * (float)(Math.PI / 180)))) * Clock.ElapsedFrameTime))
                        );
                }
                else
                {
                    // Update the position
                    note.Position += new Vector2(
                        -(NOTE_SPEED * (float)Math.Cos((NoteLaneExtension.GetAngle(tapNote.Lane) + 90f) *
                                                       (float)(Math.PI / 180))),
                        -(NOTE_SPEED * (float)Math.Sin((NoteLaneExtension.GetAngle(tapNote.Lane) + 90f) *
                                                       (float)(Math.PI / 180)))
                    );
                }
            }
        }
    }

    base.Update();
}
```

The result is very accurate but missing like from 1 ms to max like to 20 when the `ElapsedFrameTime` is high when inactive.

```csharp
// Note set time goal at 5000 ms
// Active
[runtime] 2022-03-25 22:56:48 [verbose]: Note spawn at 4500.8809
[runtime] 2022-03-25 22:56:48 [verbose]: Despawned at 5000.8792

// Inactive
[runtime] 2022-03-25 23:01:09 [verbose]: Note spawn at 4501.634
[runtime] 2022-03-25 23:01:10 [verbose]: Despawned at 5019.570500000001
```

And from this PR with new logic now we can control by slower or faster the time now. (But the note that's spawn with old logic still cannot control the time.)

### CODE REVIEW

File: maisim/maisim.Game/Screen/Gameplay/Playfield.cs [hunk 1]
```diff
@@ -42,39 +52,73 @@ public void SpawnTapNote(NoteLane lane)
...
         protected override void Update()
         {
+            if (NotesPool != null)
+            {
+                foreach (Drawable note in NotesPool.ToList())
+                {
+                    // We have a lot of note type so we need to check on its type before we can update it
+                    if (note is DrawableTapNote tapNote)
+                    {
+                        if (tapNote.TargetTime != 0f && Clock.TimeInfo.Current >= tapNote.TargetTime - TIME_NOTE_APPEARS)
```
Comment (Game4all) [inline maisim/maisim.Game/Screen/Gameplay/Playfield.cs:74]: Not sure about making a huge condition block for updating note position, sounds like something for an overridable method.

**Review by Game4all — CHANGES_REQUESTED** [review]
not sure abt the general direction

File: maisim/maisim.Game/Screen/Gameplay/Playfield.cs (hunk 1 above)
Comment (HelloYeew) [reply maisim/maisim.Game/Screen/Gameplay/Playfield.cs:74]: Will try to move it as a method some like stratrgy.

**Comment by HelloYeew** [comment]
I move the update and spawn process to a seperate method. The `TapNote` now has a separate method on update call `updateTapNotePosition`. But `spawnNoteFromPool` method I'm still not much sure on its process maybe we need to implement other note type to the spawn process to clearly see the proper way.

**Review by Game4all — COMMENTED** [review]
Seems like my suggestion wasn't clear enough, my bad.

**Comment by HelloYeew** [comment]
Now I try to fix `Playfield` is null by add the constructor to create a new `Playfield` but it stuck that the playfield is not get update because it's not inside the drawable so I just add playfield to `InternalChildren` inside load method and it's work.

File: maisim/maisim.Game/Screen/Gameplay/Playfield.cs [hunk 2]
```diff
@@ -42,40 +52,97 @@ public void SpawnTapNote(NoteLane lane)
...
+                ),
                 Lane = lane
             });
         }
 
+        /// <summary>
+        /// Update the target <see cref="DrawableNote"/>'s position.
+        /// </summary>
+        /// <param name="tapNote">The <see cref="DrawableTapNote"/> that want to update.</param>
+        private void updateTapNotePosition(DrawableTapNote tapNote)
```
[the remark this answers is not in the document]
Comment (HelloYeew) [reply maisim/maisim.Game/Screen/Gameplay/Playfield.cs]: I added `UpdatePosition` and `CanDespawn` method now it can work seperately.

**Comment by Game4all** [comment]
Just pushed a few commits that should pave the path forward, as the previously envisioned direction has severe flaws. @HelloYeew requesting review.

**Comment by HelloYeew** [comment]
@Game4all The new change that's using `Bindable` as a state is looking good but there is one problem in `GameElement` test scene since the note has been added and set as active by default so the note is moving out of the screen (Or maybe we are not need that test scene anymore?)

I pushed the fix on the wrong note origin (In the `PlayfieldScreen` is right since it's manually set the start position but in `NoteSpawnTimer` has started on the origin (0,0) since the note's origin has not been set by default) and some animation that's not the same as before refactor.

**Review by Game4all — APPROVED** [review]
let's get this is as is for now to allow experimenting a bit further.

**Pull request merged**